FRC-Extension/RoboRIOCode/RoboRIOConnection.cs

```diff
@@ -26,73 +30,189 @@ public enum ConnectionUser
+                m_teamNumber = 0;
+            }
+            string roboRioMDNS = string.Format(RoboRioMdnsFormatString, m_teamNumber);
+            string roboRioLan = string.Format(RoboRioLanFormatString, m_teamNumber);
+            string roboRIOIP = string.Format(RoboRioIpFormatString, m_teamNumber / 100, m_teamNumber % 100);
+
+            using (TcpClient usbClient = new TcpClient())
+            using (TcpClient mDnsClient = new TcpClient())
+            using (TcpClient lanClient = new TcpClient())
+            using (TcpClient ipClient = new TcpClient())
+            {
+
+                Task usb = usbClient.ConnectAsync(RoboRioUSBIp, 80);
+                Task mDns = mDnsClient.ConnectAsync(roboRioMDNS, 80);
+                Task lan = lanClient.ConnectAsync(roboRioLan, 80);
+                Task ip = ipClient.ConnectAsync(roboRIOIP, 80);
+                Task delayTask = Task.Delay(10000);
+
+                // http://stackoverflow.com/questions/24441474/await-list-of-async-predicates-but-drop-out-on-first-false
+                List<Task> tasks = new List<Task>()
+                {
+                    usb, mDns, lan, ip, delayTask
+                };
+
+                while (tasks.Count != 0)
+                {
+                    var finished = await Task.WhenAny(tasks);
+
+                    if (finished == delayTask)
+                    {
+                        return false;
+                    }
+                    else if (finished.IsCompleted && !finished.IsFaulted && !finished.IsCanceled)
+                    {
+                        // A task finished, find our host
+                        TcpClient foundHost = null;
+                        if (finished == usb)
+                        {
+                            foundHost = usbClient;
+                            m_connectionType = ConnectionType.USB;
+                        }
+                        else if (finished == mDns)
+                        {
+                            foundHost = mDnsClient;
+                            m_connectionType = ConnectionType.MDNS;
+                        }
+                        else if (finished == lan)
+                        {
+                            foundHost = lanClient;
+                            m_connectionType = ConnectionType.LAN;
+                        }
+                        else if (finished == ip)
+                        {
+                            foundHost = ipClient;
+                            m_connectionType = ConnectionType.IP;
+                        }
+                        else
+                        {
+                            // Error
+                            return false;
+                        }
+
+                        var ep = foundHost.Client.RemoteEndPoint;
+                        var ipEp = ep as IPEndPoint;
+                        if (ipEp == null)
+                        {
+                            // Continue, we cannot use this one
```

Comment:
Is there ever a realistic situation where `RemoteEndPoint` could be not an `IPEndPoint`? If not, then make this a standard cast, that way an exception will get thrown if this isn't true.